### ISSUE TITLE
GPU-resident decode ops are net-negative on Metal: they disqualify the decode replay tape (-8% to -13% tg)

### ISSUE THREAD

**Issue by digitaloten** [kryptic-sh/infr#47]
The GPU-resident decode path (`Op::Argmax`, `Op::Sample`, `Op::EmbedGather`) is **numerically correct** on Metal — validated on M3 Pro in #46, and greedy decode with the ops active is byte-identical to the same build on the host paths. This issue is purely about **performance policy**: on Metal the ops cost 8–13% of decode throughput and buy nothing.

## Measurements (M3 Pro, macOS 26.5, greedy `INFR_TEMP=0`, `bench -p 0 -n 128`)

Same binary, A/B via the built-in escapes `INFR_NO_GPU_ARGMAX=1 INFR_NO_GPU_EMBED=1`.

| model | GPU-decode ON | OFF | delta |
|---|---|---|---|
| Qwen3-0.6B-Q5_K_M (`r=10`) | 136.68 t/s | **154.76 t/s** | **-11.7%** |
| Qwen3-0.6B-Q3_K_M (`r=10`) | 161.30 t/s | **175.47 t/s** | **-8.1%** |
| Qwen3.5-4B-MTP-Q4_K_XL (`r=6`) | 31.42 t/s | **35.93 t/s** | **-12.6%** |

It does not shrink with model size — the cost is per-token and structural, not per-FLOP.

## Mechanism

`replay_shape` (`crates/infr-metal/src/exec.rs`) enumerates the ops the decode replay tape can record and ends in `_ => return false`. `Op::Argmax`, `Op::Sample` and `Op::EmbedGather` are not among the allowed arms. So **any one of them in the decode graph disqualifies the whole graph from the replay tape**, and every token re-walks and re-encodes the full op list instead of replaying the flat recorded dispatch list.

The isolation A/B matches that prediction exactly — the cost is non-additive, because *either* op alone is enough to kill the tape:

| config (0.6B-Q5_K_M, `r=5`) | tg |
|---|---|
| both on | 140.74 / 140.66 |
| `INFR_NO_GPU_ARGMAX` (embed still on) | 140.30 |
| `INFR_NO_GPU_EMBED` (argmax still on) | 139.42 |
| both off | 156.44 / 151.98 |

Meanwhile the thing these ops optimize away — the `[vocab]` logits download — is already ~free on Metal. Measured in #35: the logits readback is **0.02 ms/token** on unified memory (1 MB is a memcpy, not a bus transfer). The Vulkan economics (discrete GPU, logits cross PCIe) do not transfer.

So on Metal the trade is: give up the replay tape (a real per-token win) to avoid a readback that costs nothing.

## Secondary effect: it inflates the Metal MTP ratio

On the 4B run above, with GPU-decode **off** the autoregressive baseline rises to 35.93 t/s while `mtp_ts` stays ~35.4 (the MTP driver has its own path). The reported `mtp_ratio` therefore moves **1.116x -> 0.984x**. The Metal MTP net-positive measured earlier was in part an artifact of a depressed autoregressive baseline, not an MTP gain. Worth re-reading any Metal MTP ratio taken with the GPU decode ops active.

## Why the capability flags alone do not fix it

`embed_gather` and `gpu_sample` are backend capabilities, so Metal can opt out of those. But `gpu_argmax` in `seam/runner.rs` has **no capability gate**:

```rust
let gpu_argmax = (sampler.temp <= 0.0 || sampler.top_k == 1)
    && constraint.is_none()
    && std::env::var("INFR_NO_GPU_ARGMAX").is_err();
```

Greedy decode is the common case, so flipping Metal to `embed_gather: false, gpu_sample: false` still leaves `Op::Argmax` in the graph and the tape still never records.

## Suggested fixes (maintainer’s call — this straddles the shared runner)

1. **Opt out on Metal.** Add `gpu_argmax: bool` to `Capabilities` mirroring the existing `gpu_sample`, gate the runner on it (Vulkan `true`, Metal `false`), and set Metal `embed_gather: false`, `gpu_sample: false`. Recovers the 8–13%. Smallest change, but adds a capability whose only job is to say "no".
2. **Make them tape-recordable on Metal.** Admit the ops in `replay_shape` and record them. `Op::Argmax`/`Op::Sample` look straightforward (position-independent, fixed in/out buffers). `Op::EmbedGather` is the hard one — its `ids` Input changes per token and is currently host-mirrored, which is precisely why it was excluded. This keeps the ops and removes the regression.

Happy to implement either once you pick — (1) is a small cross-backend change; (2) is Metal-local but needs the `ids` buffer moved on-device for the tape.

Kernel correctness is not in question: #46 adds parity tests for all three (bit-exact vs the CPU interpreter across formats, sampler configs and uniform draws), and they pass.

**Comment by mxaddict** [comment]
@digitaloten "Make them tape-recordable on Metal" is a better long term solution I think.
